bundle/config/mutator/python/apply_python_mutator.go

```diff
@@ -0,0 +1,190 @@
+	return fmt.Sprintf("ApplyPythonMutator(%s)", m.phase)
+}
+
+func getExperimental(b *bundle.Bundle) config.Experimental {
+	if b.Config.Experimental == nil {
+		return config.Experimental{}
+	}
+
+	return *b.Config.Experimental
+}
+
+func (m *applyPythonMutator) Apply(ctx context.Context, b *bundle.Bundle) diag.Diagnostics {
+	experimental := getExperimental(b)
+
+	if !experimental.PyDABs.Enable {
+		log.Debugf(ctx, "'experimental.pydabs.enable' isn't enabled, skipping")
+		return nil
+	}
+
+	if experimental.PyDABs.Enable && experimental.VEnv.Path == "" {
+		return diag.Errorf("'experimental.pydabs.enable' can only be used when 'experimental.venv.path' is set")
+	}
+
+	err := b.Config.Mutate(func(leftRoot dyn.Value) (dyn.Value, error) {
+		pythonPath := path.Join(experimental.VEnv.Path, "bin", "python3")
+
+		if _, err := os.Stat(pythonPath); err != nil {
+			if os.IsNotExist(err) {
+				return dyn.InvalidValue, fmt.Errorf("can't find '%s', check if venv is created", pythonPath)
+			} else {
+				return dyn.InvalidValue, fmt.Errorf("can't find '%s': %w", pythonPath, err)
+			}
+		}
+
+		rightRoot, err := m.runPythonMutator(ctx, b.RootPath, pythonPath, leftRoot)
+
+		if err != nil {
+			return dyn.InvalidValue, err
+		}
+
+		return merge.Override(leftRoot, rightRoot, createOverrideVisitor(ctx, m.phase))
+	})
+
+	return diag.FromErr(err)
+}
+
+func (m *applyPythonMutator) runPythonMutator(ctx context.Context, rootPath string, pythonPath string, root dyn.Value) (dyn.Value, error) {
+	args := []string{
+		pythonPath,
+		"-m",
+		"databricks.bundles.build",
+		"--phase",
+		string(m.phase),
+	}
+
+	// we need to marshal dyn.Value instead of bundle.Config to JSON to support
+	// non-string fields assigned with bundle variables
+	rootConfigJson, err := json.Marshal(root.AsAny())
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to marshal root config: %w", err)
+	}
+
+	logWriter := newLogWriter(ctx, "stderr: ")
+
+	stdout, err := process.Background(
+		ctx,
+		args,
+		process.WithDir(rootPath),
+		process.WithStderrWriter(logWriter),
+		process.WithStdinReader(bytes.NewBuffer(rootConfigJson)),
+	)
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("python mutator process failed: %w", err)
+	}
+
+	// we need absolute path, or because later parts of pipeline assume all paths are absolute
+	// and this file will be used as location
+	virtualPath, err := filepath.Abs(filepath.Join(rootPath, "__generated__.yml"))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to get absolute path: %w", err)
+	}
+
+	generated, err := yamlloader.LoadYAML(virtualPath, bytes.NewReader([]byte(stdout)))
+
+	if err != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to parse Python mutator output: %w", err)
+	}
+
+	normalized, diagnostic := convert.Normalize(config.Root{}, generated)
+
+	if diagnostic.Error() != nil {
+		return dyn.InvalidValue, fmt.Errorf("failed to normalize Python mutator output: %w", diagnostic.Error())
+	}
```

Comment:
What about the warnings? If PyDABs produces invalid configuration (e.g. unrecognized keys or values whose type doesn't match the expected type), then I think it should really fail.

Reply:
Good point, I will also add a test for it